communication/src/protocol_defs.h

```diff
@@ -18,20 +18,33 @@ namespace particle { namespace protocol {
 
 enum ProtocolError
 {
-	NO_ERROR,				// 0
-	PING_TIMEOUT,			// 1
-	IO_ERROR,				// 2
-	INVALID_STATE,			// 3
-	INSUFFICIENT_STORAGE,	// 4
-	MALFORMED_MESSAGE,		// 5
-	DECRYPTION_ERROR,		// 6
-	ENCRYPTION_ERROR,		// 7
-	AUTHENTICATION_ERROR,	// 8
-	BANDWIDTH_EXCEEDED,		// 9
-	MESSAGE_TIMEOUT,			// 10
-	MISSING_MESSAGE_ID,		// 11
-	MESSAGE_RESET,			// 12
-	SESSION_RESUMED,			// 13
+	/* 00 */ NO_ERROR,
+	/* 01 */ PING_TIMEOUT,
+	/* 02 */ IO_ERROR,	// too generic, discontinue using this.  Perfer/add a specific one below
+	/* 03 */ INVALID_STATE,
+	/* 04 */ INSUFFICIENT_STORAGE,
+	/* 05 */ MALFORMED_MESSAGE,
+	/* 06 */ DECRYPTION_ERROR,
+	/* 07 */ ENCRYPTION_ERROR,
+	/* 08 */ AUTHENTICATION_ERROR,
+	/* 09 */ BANDWIDTH_EXCEEDED,
+	/* 10 */ MESSAGE_TIMEOUT,
+	/* 11 */ MISSING_MESSAGE_ID,
+	/* 12 */ MESSAGE_RESET,
+	/* 13 */ SESSION_RESUMED,
+	/* 14 */ IO_ERROR_FORWARD_MESSAGE_CHANNEL,
+	/* 15 */ IO_ERROR_SET_DATA_MAX_EXCEEDED,
+	/* 16 */ IO_ERROR_PARSING_SERVER_PUBLIC_KEY,
+	/* 17 */ IO_ERROR_GENERIC_ESTABLISH,
+	/* 18 */ IO_ERROR_GENERIC_RECEIVE,
+	/* 19 */ IO_ERROR_GENERIC_SEND,
+	/* 20 */ IO_ERROR_GENERIC_MBEDTLS_SSL_WRITE,
+	/* 21 */ IO_ERROR_DISCARD_SESSION,
+	/* 21 */ IO_ERROR_LIGHTSSL_BLOCKING_SEND,
+	/* 22 */ IO_ERROR_LIGHTSSL_BLOCKING_RECEIVE,
+	/* 23 */ IO_ERROR_LIGHTSSL_RECEIVE,
+	/* 24 */ IO_ERROR_LIGHTSSL_HANDSHAKE_NONCE,
+	/* 25 */ IO_ERROR_LIGHTSSL_HANDSHAKE_RECV_KEY,
```

Comment:
Thanks for this!  I'll add that with a note that if the ProtocolError's get expanded they also need to be added to the `system_error particle::protocol::toSystemError(ProtocolError error) {` switch.